### PR TITLE
CMake's better handling c++14 requirement

### PR DESCRIPTION
**Pull request by veblush**
As Cmake got a better way to specify that C++14 or above is required with [target_compile_features](https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html#id5), gRPC needs to use this instead of using `CMAKE_CXX_STANDARD` which should be done by the application. This is also aligned with [how Cloud C++ does](https://github.com/googleapis/google-cloud-cpp/blob/5a32dd780d87a2af5b5df6f014c37ba600f8a8fb/cmake/FindgRPC.cmake#L260).

In order for gRPC to get this change, the minium version of cmake should be bumped to 3.8, which is acceptable as OSS C++ policy needs 3.10 as the oldest version of cmake to support.

### CODE REVIEW

**Comment by veblush** [comment]
CC @coryan

File: examples/android/helloworld/app/CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.4.1)
+cmake_minimum_required(VERSION 3.8)
 
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
```
Comment (coryan) [inline examples/android/helloworld/app/CMakeLists.txt:3]: Drive-by: looks like yo could remove this?

File: examples/cpp/cmake/common.cmake [hunk 2]
```diff
@@ -17,7 +17,7 @@
 # See cmake_externalproject/CMakeLists.txt for all-in-one cmake build
 # that automatically builds all the dependencies before building route_guide.
 
-cmake_minimum_required(VERSION 3.5.1)
+cmake_minimum_required(VERSION 3.8)
 
 if (NOT DEFINED CMAKE_CXX_STANDARD)
   set (CMAKE_CXX_STANDARD 14)
```
Comment (coryan) [inline examples/cpp/cmake/common.cmake:23]: Seems like you could remove this too?

File: src/android/test/interop/app/CMakeLists.txt [hunk 3]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.4.1)
+cmake_minimum_required(VERSION 3.8)
 
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
```
Comment (coryan) [inline src/android/test/interop/app/CMakeLists.txt:3]: Maybe you can remove this?

File: examples/android/helloworld/app/CMakeLists.txt (hunk 1 above)
Comment (veblush) [reply examples/android/helloworld/app/CMakeLists.txt:3]: Is this because gRPC cmakefile has this already?

Comment (coryan) [reply examples/android/helloworld/app/CMakeLists.txt:3]: Yes, any target that depends on gRPC++ will inherit the requirement to use C++ 14 or higher.

File: examples/cpp/cmake/common.cmake (hunk 2 above)
Comment (veblush) [reply examples/cpp/cmake/common.cmake]: Done.

File: src/android/test/interop/app/CMakeLists.txt (hunk 3 above)
Comment (veblush) [reply src/android/test/interop/app/CMakeLists.txt]: Done.

**Comment by veblush** [comment]
@drfloob AJ, would you review this as Jan is OOO?

File: tools/run_tests/run_tests.py [hunk 4]
```diff
@@ -477,7 +477,7 @@ def _compiler_options(self, use_docker, compiler):
             _check_compiler(compiler, ['default', 'cmake'])
 
         if compiler == 'default' or compiler == 'cmake':
-            return ('debian11', [])
+            return ('debian11', ["-DCMAKE_CXX_STANDARD=14"])
```
Comment (jtattermusch) [inline tools/run_tests/run_tests.py:480]: It feels odd that to our "default" build via run_tests.py we need to pass extra options.
Would that mean that  if one builds with cmake by hand, without passing any extra configuration, they would get a differently configured build that if they run run_tests.py? If so, that feels wrong.

**Review by jtattermusch — COMMENTED** [review]
Overall looks fine, but I had one concern - see comment.

File: tools/run_tests/run_tests.py (hunk 4 above)
Comment (veblush) [reply tools/run_tests/run_tests.py:480]: This is required because of Apple clang which uses C++98 by default (yay). Also this is only for tests and gRPC library targets don't specify it so that C++ version can be picked by applications.

**Review by drfloob — APPROVED** [review]
